flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java

```diff
@@ -329,74 +332,164 @@ private void restoreWithRescaling(Collection<KeyedStateHandle> restoreStateHandl
-                                    break;
-                                }
-
-                                iterator.next();
-                            }
-                        } // releases native iterator resources
+                    // Clip all tmp db to Range [startKeyGroupPrefixBytes, stopKeyGroupPrefixBytes)
+                    RocksDBIncrementalCheckpointUtils.clipColumnFamilies(
+                            tmpRestoreDBInfo.db,
+                            tmpColumnFamilyHandles,
+                            startKeyGroupPrefixBytes,
+                            stopKeyGroupPrefixBytes);
+
+                    // Export all the Column Families
+                    List<Pair<RegisteredStateMetaInfoBase, ExportImportFilesMetaData>>
+                            exportedCFAndMetaData =
+                                    RocksDBIncrementalCheckpointUtils.exportColumnFamilies(
+                                            tmpRestoreDBInfo.db,
+                                            tmpColumnFamilyHandles,
+                                            tmpRestoreDBInfo.stateMetaInfoSnapshots,
+                                            exportCfBasePath);
+
+                    for (Pair<RegisteredStateMetaInfoBase, ExportImportFilesMetaData> entry :
+                            exportedCFAndMetaData) {
+                        ExportImportFilesMetaData cfMetaData = entry.getValue();
+                        // TODO: method files() doesn't exist in the RocksDB API
+                        //                        if (cfMetaData.files().isEmpty()) {
+                        //                            continue;
+                        //                        }
+                        columnFamilyMetaDataToImport
+                                .computeIfAbsent(entry.getKey(), (k) -> new ArrayList<>())
+                                .add(cfMetaData);
                     }
-                    logger.info(
-                            "Finished restoring from state handle: {} with rescaling.",
-                            stateHandle);
                 }
+                logger.info(
+                        "Finished exporting column family from state handle: {} for rescaling.",
+                        stateHandle);
             }
+
+            // Open the target RocksDB and import the exported column families
+            this.rocksHandle.openDB();
+            columnFamilyMetaDataToImport.forEach(
+                    this.rocksHandle::registerStateColumnFamilyHandleWithImport);
+            logger.info(
+                    "Finished importing exported column families into target DB for rescaling.");
         } finally {
-            // Cleanup all download directories
-            allDownloadSpecs.stream()
-                    .map(StateHandleDownloadSpec::getDownloadDestination)
-                    .forEach(this::cleanUpPathQuietly);
+            cleanUpPathQuietly(exportCfBasePath);
```

Comment:
Just to clarify, `cleanUpPathQuietly(exportCfBasePath)` doesn't affect the opened RocksDb instance by deleting the imported files?

Reply:
No, I think the files are hard-linked in the directory of the created instance to which we import.